### ISSUE TITLE
borg vs borg2 - Name should change

### ISSUE THREAD

**Issue by dagelf** [borgbackup/borg#7140]
If an application makes breaking changes, its best to change the name. Reason being that a lot of dependent automations and wrappers will break. I propose `borg2`. Both old and new can be shipped in the same package. Borg2 can eventually replace borg, but just with a wrapper that makes the command line interface compatible so that old wrappers can continue working.

**Comment by ThomasWaldmann** [comment]
Yes, I agree that the borg 2.x cli name should be `borg2` until some admin decides otherwise. So both borg versions can be run in parallel until the migration is completed.

Not sure if distributions should also switch their borg 1.x packages to a borg1 cli command name, but guess that would be best so one can have a symlink `borg -> borg1` or `borg -> borg2` and also directly invoke `borg1` or `borg2`.

The remaining question is the mechanism implementing that cli command name and who implements that.

Here i think this should be a distribution-specific mechanism like the "alternatives" system of debian/ubuntu or a manually managed symlink. Thus, either the dist package maintainer or the sysadmin would implement this.

For the software and documentation in the borg github repo and the python pypi package, I would like to keep the "borg" cli command name and also the `borgbackup` pypi package name.

**Comment by dagelf** [comment]
> So, if we merge these, next release from master will not be 1.3, but 2.0.

Changing the module name from `borg` to `borg2` at this point is something to be thoroughly considered.

Both, to eventually play with potential (meanwhile obsoleted already) export/import tar magic, but also to be able to test 1.2 in parallel with 1.1 in production across all my systems in a sane manner, I went on the surprisingly painful adventure to create myself a variant of the distribution's package that can be installed and used in parallel with the stock 1.1 one. In a hackish manner, one could install borg below a different path, but that is nothing any distribution would do, I went the painful way to do such a rename in there.
(IOW happy to clean that up and even break out some of the cases where absolute imports were used without need and against the common practice in most other similar places in the code).

For the original idea of export-import migration this would be a requirement, here it is not, but in practice, for people backing up to multiple repos, scenarios like migrating the local one to 2.0 while still waiting an undefined time for the borg storage provider the external one resides on to support 2.0 could be very common.

_Originally posted by @elho in https://github.com/borgbackup/borg/issues/6602#issuecomment-1120203448_

**Comment by ThomasWaldmann** [comment]
Our docs are versioned and automatically created from tags in the github repo (see the version selector at the lower right), so users are expected to read the docs version matching their borg version.

They don't always do that (or do not read the docs at all), but that is not our fault. :-)

**Comment by ThomasWaldmann** [comment]
To keep archives for longer, users are advised to create new borg 2 repos and then use borg transfer to copy all their existing archives into it. That avoids the issue that you would have to use an old (and by then unsupported) borg 1 version to access backups in some years in the future.

**Comment by dagelf** [comment]
The problem is that many people create archives and then forget about it. They might only get the message to migrate 10 years from now.

**Comment by ThomasWaldmann** [comment]
Guess they would notice by the time when dists start to remove the borg 1.x package (see the python 2 to 3 migration).

**Comment by dagelf** [comment]
The Python migration was the biggest disaster in modern software IMHO. We should make technology easier to use and more inclusive!

**Comment by deathtrip** [comment]
And what about the wasted man-hours spent on maintaining the backward compatibility you propose?
If someone finds their old archives are no longer supported, they can download an old borg 1.x version and still access those.

**Comment by dagelf** [comment]
The major issue at hand is the new cli - as there are numerous frontends and automations that will break in all kinds of colourful ways. Hence my proposal to keep the new name `borg2` in perpetuity.

> And what about the wasted man-hours spent on maintaining the backward compatibility you propose? If someone finds their old archives are no longer supported, they can download an old borg 1.x version and still access those.

I think this is very much on topic. 

My proposal is to side step this issue by not maintaining backward compatibility (assuming it is a burden) but instead just stick to a new name `borg2`. That way `borg` still works as it always has, even if abandoned. Something that is not clear to me is whether the old binary can be packaged along with the new one, in the same package... that way no new packaging needs to be done anywhere. The only potential downside I see is that if a security issue does crop up in the old one, it might affect distribution of the new one, as they would be in the same package.

**Comment by dagelf** [comment]
To add, it leaves the option open to upgrade `borg` with a wrapper to `borg2` - that maintains the old cli. Of course that will work best if `borg2` can continue supporting the current algorithms. But it doesn't force it, it just leaves the door open.

**Comment by ThomasWaldmann** [comment]
I don't think anyone would like to package borg 1.x into the same package as borg 2. That is just bad for maintenance and updates and just creates additional complexity and work compared to having 2 packages that live side-by-side and can be updated independently.

Guess this applies to (linux and other) dist packages as well as to the fat-binary packages provided by us.

BTW, the whole point of making a (breaking) major release is getting rid of legacy code, but there is another ticket that already described why that path was chosen.

**Comment by dagelf** [comment]
The other discussion, linked above, seems to center on two issues:
1) Breaking CLI changes
2) Dropping legacy crypto code

If the CLI stays the same, there should be no need for this ticket's whole discussion. So that brings up the question - how about that if `borg` detects `::` in the CLI, then it uses the old format - otherwise the new. Ie. is it possible to make the cli trivially backwards compatible?  Ie. no breaking changes in CLI - but do extend it with a new CLI format? My suggestion is just do this with a wrapper, if you want to get completely rid of old code in the main binary. 

As for introducing borg2 as a new package, it is easier to maintain one package than two...  and we don't want people to keep installing the old version - hence the package name should stay `borg`. And it's justifiable if there's a roadmap to a merge. That way we can keep just one package `borg` - and as you initially suggested. If non-backwards-compatible CLI is a must,  have `borg2` as a command in it, and later drop the `borg2` for `borg`, still in the same package - and If someone can't decompress a repo, they can just install an old version via pip, and that can be maintained and mentioned in the documentation. As for newer distributions and their packaging, you can expect them to not have old versions available anymore at some point.

People who only get introduced to borg now, won't have any confusion because new defaults can ensure that.

**Comment by ThomasWaldmann** [comment]
@dagelf You're late to the party: the old cli argparsing code is already gone and won't come back. the old crypto code is still there for migration reasons, but will be gone in 2.1.

Also, sorry for being repetitive: the point of the borg 2 breaking release is to get rid of a lot of legacy code, that includes the legacy cli argparsing. I won't write a cli-compatibility wrapper.

**Comment by ThomasWaldmann** [comment]
Also, I am closing this ticket now as it seems to be impossible to stay on-topic.
